NBitcoin/Scripting/Parser/IInput.cs

```diff
@@ -0,0 +1,18 @@
+using System;
+using System.Collections.Generic;
+
+namespace NBitcoin.Scripting.Parser
+{
+    public interface IInput<out T>
```

Comment:
This one should be internal ?

Reply:
Also, I would suggest to inherit  from `IEnumerator<out T>` and remove `Advance/GetCurrent/GetNext`.
Then just make methods returning `IInput<out T>` before implement `IEnumerable<T>` and return `IInput<T>` .

The nice thing is that people can use `foreach` then.

Reply:
Nice. I wonder why this didn't come up to my mind...

Reply:
I've been trying to inherit from IEnumerator<out T>. But it is not working well :(
It is because that Input for Parser combinator must be immutable. (`Advance()` will create another instance each time it proceeds)
IEnumerator is only for mutable object.
I think inheriting from `IEnumerable` and delegate GetEnuemrator call to underlying Source object is enough.